### ISSUE TITLE
Can't resolve lodash/throttle

### ISSUE THREAD

**Issue by 01binary** [STORIS/react-scrollbar-size#6]
I tried installing and building many times, but I can't get the webpack2 build to succeed:
```
ERROR in ../node_modules/react-scrollbar-size/ScrollbarSize.js
Module not found: Error: Can't resolve 'lodash/throttle' in '..\node_modules\react-scrollbar-size'
 @ ../node_modules/react-scrollbar-size/ScrollbarSize.js 39:16-42
 @ ../node_modules/react-scrollbar-size/index.js
 @ ../node_modules/material-ui/Tabs/Tabs.js
 @ ../node_modules/material-ui/Tabs/index.js
 @ ../node_modules/material-ui/index.es.js
 @ ./components/Welcome.js
```
I verified that NPM installs lodash into node_modules and it does have throttle exported from index.js. 

Using Node 8 LTS (8.9.3) and NPM 5.5.1. This is probably some configuration mistake, and I will post here if I manage to solve it - might be useful to keep this issue around in case someone else has the same problem.

**Comment by 01binary** [comment]
`npm i` returns

```
npm WARN react-scrollbar-size@2.0.2 requires a peer of lodash@^4.17.0 but none is installed. You must install peer dependencies yourself.
```

`npm list lodash` returns

```
+-- babel-core@6.26.0
| +-- babel-generator@6.26.0
| | `-- lodash@4.17.4
| +-- babel-register@6.26.0
| | `-- lodash@4.17.4
| +-- babel-template@6.26.0
| | `-- lodash@4.17.4
| +-- babel-traverse@6.26.0
| | `-- lodash@4.17.4
| +-- babel-types@6.26.0
| | `-- lodash@4.17.4
| `-- lodash@4.17.4
+-- babel-preset-env@1.6.1
| +-- babel-plugin-transform-es2015-block-scoping@6.26.0
| | `-- lodash@4.17.4
| +-- babel-plugin-transform-es2015-classes@6.24.1
| | `-- babel-helper-define-map@6.26.0
| |   `-- lodash@4.17.4
| `-- babel-plugin-transform-es2015-sticky-regex@6.24.1
|   `-- babel-helper-regex@6.26.0
|     `-- lodash@4.17.4
+-- cheerio@0.19.0
| `-- lodash@3.10.1
+-- copy-webpack-plugin@4.3.1
| `-- lodash@4.17.4
+-- enzyme@3.1.1
| +-- cheerio@1.0.0-rc.2
| | `-- lodash@4.17.4  deduped
| `-- lodash@4.17.4
+-- enzyme-adapter-react-16@1.0.4
| +-- enzyme-adapter-utils@1.3.0
| | `-- lodash@4.17.4
| `-- lodash@4.17.4
+-- enzyme-to-json@3.2.2
| `-- lodash@4.17.4
+-- eslint@4.13.1
| +-- inquirer@3.3.0
| | `-- lodash@4.17.4
| +-- lodash@4.17.4
| `-- table@4.0.2
|   `-- lodash@4.17.4
+-- jest@21.2.1
| `-- jest-cli@21.2.1
|   `-- istanbul-api@1.2.1
|     `-- async@2.6.0
|       `-- lodash@4.17.4
+-- webpack@3.10.0
| +-- async@2.6.0
| | `-- lodash@4.17.4
| `-- watchpack@1.4.0
|   `-- async@2.6.0
|     `-- lodash@4.17.4
+-- webpack-bundle-analyzer@2.9.1
| `-- lodash@4.17.4
`-- whs-structured-logger@1.3.3
  `-- bunyan-rotating-file-stream@1.6.3
    `-- lodash@4.17.4
```

`npm list react-scrollbar-size` returns

```
ui-tank@1.0.0 C:\Users\vnovytskyy\Desktop\tank
`-- material-ui@1.0.0-beta.24
  `-- react-scrollbar-size@2.0.2
```

It wants ^4.17.0 and I have 4.17.4 - so far I don't understand why there's a problem

**Comment by 01binary** [comment]
Detailed webpack output. Starting to look like version mismatch.

```
ERROR in ../node_modules/react-scrollbar-size/ScrollbarSize.js
Module not found: Error: Can't resolve 'lodash/throttle' in 'C:\myapp\node_modules\react-scrollbar-size'
resolve 'lodash/throttle' in 'C:\myapp\node_modules\react-scrollbar-size'
  Parsed request is a module
  using description file: C:\myapp\node_modules\react-scrollbar-size\package.json (relative path: .)
    Field 'browser' doesn't contain a valid alias configuration
  after using description file: C:\myapp\node_modules\react-scrollbar-size\package.json (relative path: .)
    resolve as module
      C:\myapp\node_modules\react-scrollbar-size\node_modules doesn't exist or is not a directory
      C:\myapp\node_modules\node_modules doesn't exist or is not a directory
      C:\node_modules doesn't exist or is not a directory
      C:\Users\vnovytskyy\node_modules doesn't exist or is not a directory
      C:\Users\node_modules doesn't exist or is not a directory
      C:\node_modules doesn't exist or is not a directory
      looking for modules in C:\myapp\node_modules
        using description file: C:\myapp\package.json (relative path: ./node_modules)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\myapp\package.json (relative path: ./node_modules)
          using description file: C:\myapp\node_modules\lodash\package.json (relative path: ./throttle)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\myapp\node_modules\lodash\throttle doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\myapp\node_modules\lodash\throttle.js doesn't exist
            .json
              Field 'browser' doesn't contain a valid alias configuration
              C:\myapp\node_modules\lodash\throttle.json doesn't exist
            .svg
              Field 'browser' doesn't contain a valid alias configuration
              C:\myapp\node_modules\lodash\throttle.svg doesn't exist
            as directory
              C:\myapp\node_modules\lodash\throttle doesn't exist
[C:\myapp\node_modules\react-scrollbar-size\node_modules]
[C:\myapp\node_modules\node_modules]
[C:\node_modules]
[C:\Users\vnovytskyy\node_modules]
[C:\Users\node_modules]
[C:\node_modules]
[C:\myapp\node_modules\lodash\throttle]
[C:\myapp\node_modules\lodash\throttle.js]
[C:\myapp\node_modules\lodash\throttle.json]
[C:\myapp\node_modules\lodash\throttle.svg]
[C:\myapp\node_modules\lodash\throttle]
 @ ../node_modules/react-scrollbar-size/ScrollbarSize.js 39:16-42
 @ ../node_modules/react-scrollbar-size/index.js
 @ ../node_modules/material-ui/Tabs/Tabs.js
 @ ../node_modules/material-ui/Tabs/index.js
 @ ../node_modules/material-ui/index.es.js
 @ ./components/Welcome.js
 @ ./components/App.js
 @ ./client/main.js
 @ multi whatwg-fetch ./client/main.js
```

**Comment by 01binary** [comment]
Found the problem:
https://github.com/STORIS/react-scrollbar-size/blob/master/src/ScrollbarSize.jsx#L4

```
import throttle from 'lodash/throttle';
```

should be

```
import throttle from 'lodash/function/throttle';
```

...using the current version of Lodash, which apparently is the version I am forced to use if I am also using babel and material UI.

**Comment by shawnmcknight** [comment]
Sorry, I'm not really sure what problem you (were) having.  I initialized a new repo, and installed `material-ui@next`, `react`, and `react-dom`.  `lodash` installed into `project_root/node_modules/lodash` and contains `throttle.js` which is what `react-scrollbar-size` is importing.  It's not importing from lodash's `index.js` - it should directly import from the `throttle.js` file.

I didn't spin up a new webpack config and I'm not transpiling, but I'm able to successfully run a program in node which requires that module properly:
```js
const throttle = require( 'lodash/throttle');
setInterval(throttle(() => {
	console.log(new Date().toISOString());
}, 3000), 1000);
```
Executing logs a timestamp every 3 seconds which would be the expected behavior.

In your webpack output, I would have expected it to resolve the module at this line:
`[C:\myapp\node_modules\lodash\throttle.js]`

**Comment by 01binary** [comment]
Hello,

I checked again, and `lodash` folder in my `node_modules` does not contain `throttle.js`:

![image](https://user-images.githubusercontent.com/19985964/34312424-ed605f32-e718-11e7-9a63-c805b83e62e8.png)

However, `throttle.js` is inside of `function` folder:

![image](https://user-images.githubusercontent.com/19985964/34312448-136c178e-e719-11e7-9c73-706bef9d4530.png)

Here's the `package.json` from `node_modules/lodash`, whatever gets pulled in by babel and material-ui:

```
{
  "_from": "lodash@^3.2.0",
  "_id": "lodash@3.10.1",
  "_inBundle": false,
  "_integrity": "sha1-W/Rejkm6QYnhfUgnid/RW9FAt7Y=",
  "_location": "/lodash",
  "_phantomChildren": {},
  "_requested": {
    "type": "range",
    "registry": true,
    "raw": "lodash@^3.2.0",
    "name": "lodash",
    "escapedName": "lodash",
    "rawSpec": "^3.2.0",
    "saveSpec": null,
    "fetchSpec": "^3.2.0"
  },
  "_requiredBy": [
    "/cheerio"
  ],
  "_resolved": "https://proget.dev.webmd.com/npm/npm/lodash/-/lodash-3.10.1.tgz",
  "_shasum": "5bf45e8e49ba4189e17d482789dfd15bd140b7b6",
  "_spec": "lodash@^3.2.0",
```

After I ran `npm upgrade material-ui` the problem went away, but my package manifest remained unchanged so I can't check in this fix. Looks like troubleshooting continues.

**Comment by shawnmcknight** [comment]
It looks to me like you moved off the alpha channel of material-ui.  The non-alpha release doesn't contain react-scrollbar-size and likely wouldn't have the same problems.

Have you tried deleting node_modules and doing a full `npm install` from scratch?  Your lodash folder looks incorrect and that version spec in the package.json is indicating `3.10.1` rather than the required `^4.17.0`.

If you're not making use of package locking, you might want to consider deleting your `package-lock.json` file before doing the reinstall if it exists as well.  I've had numerous problems with the npm 5.x.x releases and their handling of package locks.  We don't even use them in our projects (although we'd like to) because of issues.

**Comment by 01binary** [comment]
Thanks, I will continue troubleshooting. I thought it was likely the issue wasn't with this package, but I might be stuck for now (already tried deleting package lock and node modules folder). For some reason it refuses to install the right version of lodash.

**Comment by shawnmcknight** [comment]
Looks like the lodash version you are seeing is coming from `cheerio@0.19.0` based on your `npm list` output, although I would have expected it to install the `4.17.4` version as well to support your install of `material-ui@1.0.0-beta.24` which has `4.17.4` as a dependency  `4.17.4` should be installed somewhere that it can be accessed by `react-scrollbar-size` (either in its node_modules folder or in a parent's node_modules folder) because of the `material-ui` dependency.

Could be an npm issue with not installing multiple versions properly or could be an issue with the webpack resolution.  I'm happy to offer any assistance I can to you, but I'm pretty sure it's not an issue with this repo.

**Comment by shawnmcknight** [comment]
Out of curiosity, I installed cheerio@0.19.0 into my repo.  The end result was `node_modules/lodash` matching `4.17.4` and `node_modules/cheerio/node_modules/lodash` matching `3.10.1`.  That structure matches my expectations as cheerio can now access the version of lodash it needs, and material-ui/react-scrollbar-size can access the version of lodash they need.  Let me know if I can provide anything further.  Good luck!

**Comment by shawnmcknight** [comment]
Closing this out for now.  If you end up thinking there is an issue with this repo, please feel free to reopen and I'll have a look.

**Comment by 01binary** [comment]
No need to re-open but I think I got it to work:

- Uninstall Cheerio. The initial decision to install Cheerio as an explicit dependency is what caused all this.
- Uninstall material-ui
- Reinstall material-ui

This caused the right version of Lodash to make it into node_modules. When we installed Cheerio we were on an holder NPM version that could not lock Cheerio because it was a dependency-of-a-dependency (or something like that) so we forced the right version of Cheerio for Enzyme by installing it explicitly. Several months later, that decision caused this issue.
